source/common/router/delegating_route_impl.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "envoy/router/router.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Implementation of DelegatingRoute that delegates all method calls to the RouteConstSharedPtr
+ * base route it wraps around.
+ *
+ * Intended to be used as a route mutability mechanism, where a filter can create a derived class of
+ * DelegatingRoute and override specific methods (e.g. routeEntry) while preserving the rest of the
+ * properties/behavior of the base route.
+ */
+class DelegatingRoute : public Router::Route {
+public:
+  DelegatingRoute(Router::RouteConstSharedPtr route) : base_route_(route) {}
```

Comment:
make single arg ctors `explicit`

Reply:
Related to your comment here (https://github.com/envoyproxy/envoy/pull/15266/files#r586565475) about dereferencing the pointer without nullptr checks, I notice I do that with this `DelegatingRoute` class as well (oops).

I plan on adding in nullptr checks, since I think the cleaner implementation is to pass `RouteConstSharedPtr` by value instead of reference. For example, then callers (filter callbacks) don't need to worry about passing in a reference or checking if the `decoder_callbacks_->route()` is nullptr.

Should `DelegatingRoute` propagate nullptr is nullptr is detected? Ex: `base_route_` is nullptr, then all methods such as `routeEntry()` and `tracingConfig` also return nullptr

Reply:
>For example, then callers (filter callbacks) don't need to worry about passing in a reference or checking if the decoder_callbacks_->route() is nullptr.
route() can already return nullptr so callers already handle it. Because of this I think it makes more sense to avoid wrapping a nullptr and have the delegating route assume that its wrapping a non-null route. If we need to pass
a shared ptr to have the delegated route participate in ownership we can `ASSERT(route_ != nullptr)` in the ctor
and skip all the null checks. The suggestion to use const ref over raw-pointer makes sense if we're dealing with
a non-owning pointer, but if we need a shared pointer then an ASSERT is the best we can do.

Reply:
Thanks for the explanation, makes a lot of sense. I added an `ASSERT(base_route_ != nullptr);` to the DelegatingRoute ctor so it too can participate in ownership (will push up my branch commits soon).